### PR TITLE
Update CBMC proof harnesses

### PR DESCRIPTION
**Pull request by SaswatPadhi**
This PR fixes the 12 broken proofs that were disabled in #679.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/ring_buffer.c [hunk 1]
```diff
@@ -232,8 +232,8 @@ static inline bool s_buf_belongs_to_pool(const struct aws_ring_buffer *ring_buff
 #ifdef CBMC
     /* only continue if buf points-into ring_buffer because comparison of pointers to different objects is undefined
      * (C11 6.5.8) */
-    if ((__CPROVER_POINTER_OBJECT(buf->buffer) != __CPROVER_POINTER_OBJECT(ring_buffer->allocation)) ||
-        (__CPROVER_POINTER_OBJECT(buf->buffer) != __CPROVER_POINTER_OBJECT(ring_buffer->allocation_end))) {
+    if (!__CPROVER_same_object(buf->buffer, ring_buffer->allocation) ||
+        !__CPROVER_same_object(buf->buffer, ring_buffer->allocation_end - 1)) {
```
Comment (feliperodri) [inline source/ring_buffer.c:236]: Why do you need `ring_buffer->allocation_end - 1` here?

File: verification/cbmc/proofs/aws_byte_buf_write_be16/cbmc-batch.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+build_memory: 32000
+cbmcflags: "--unwind;1;--unwindset;memcpy_impl.0:11;--object-bits;8;--pointer-primitive-check;--bounds-check;--conversion-check;--div-by-zero-check;--float-overflow-check;--nan-check;--pointer-check;--pointer-overflow-check;--signed-overflow-check;--undefined-shift-check;--unsigned-overflow-check;--unwinding-assertions"
+coverage_memory: 32000
+expected: SUCCESSFUL
+goto: /Users/saspadhi/Repos/aws-c-common/verification/cbmc/proofs/aws_byte_buf_write_be16/gotos/aws_byte_buf_write_be16_harness.c.goto
+jobos: ubuntu16
+property_memory: 32000
+report_memory: 32000
```
Comment (feliperodri) [inline verification/cbmc/proofs/aws_byte_buf_write_be16/cbmc-batch.yaml:8]: @karkhaz Isn't these files generated automatically now?

File: verification/cbmc/stubs/memcmp_override.c [hunk 3]
```diff
@@ -10,8 +10,8 @@
 int memcmp(const void *s1, const void *s2, size_t n) {
     //  __CPROVER_HIDE:;
     int res = 0;
-    __CPROVER_precondition(__CPROVER_r_ok(s1, n), "memcmp region 1 readable");
-    __CPROVER_precondition(__CPROVER_r_ok(s2, n), "memcpy region 2 readable");
+    __CPROVER_precondition(s1 != NULL && (n == 0 || __CPROVER_r_ok(s1, n)), "memcmp region 1 readable");
+    __CPROVER_precondition(s2 != NULL && (n == 0 || __CPROVER_r_ok(s2, n)), "memcpy region 2 readable");
```
Comment (feliperodri) [inline verification/cbmc/stubs/memcmp_override.c:14]: Could you add a comment explaining why do you need to add the extra constrains?

File: source/ring_buffer.c (hunk 1 above)
Comment (SaswatPadhi) [reply source/ring_buffer.c:236]: CBMC was complaining that `ring_buffer->allocation_end` is neither null nor a valid object.

Basically, we should be comparing the buffer (which is a valid object) against `allocation + 0` to `(allocation + size - 1)` which is `allocation_end - 1`.

File: verification/cbmc/proofs/aws_byte_buf_write_be16/cbmc-batch.yaml (hunk 2 above)
Comment (karkhaz) [reply verification/cbmc/proofs/aws_byte_buf_write_be16/cbmc-batch.yaml:8]: As mentioned offline, I recommend to commit syntactically-incorrect cbmc-batch.yaml files.

File: verification/cbmc/proofs/Makefile-project-defines [hunk 4]
```diff
@@ -39,8 +39,7 @@ DEFINES += -DAWS_DEEP_CHECKS=$(AWS_DEEP_CHECKS)
 
 
 # Extra CBMC flags not enabled by Makefile.common
-CHECKFLAGS += --enum-range-check
-CHECKFLAGS += --pointer-primitive-check
```
Comment (karkhaz) [inline verification/cbmc/proofs/Makefile-project-defines:43]: I request that @danielsn approves this line before we merge this.

**Comment by karkhaz** [comment]
@markrtuttle the template submodule was advanced in #679, which is already merged

File: verification/cbmc/proofs/Makefile-project-defines (hunk 4 above)
Comment (SaswatPadhi) [reply verification/cbmc/proofs/Makefile-project-defines:43]: The changes I made in this PR fix all the enum-range-check issues as well, so we don't _need_ to disable it, but @feliperodri suggested that we do

Comment (karkhaz) [reply verification/cbmc/proofs/Makefile-project-defines:43]: I don't mind about --enum-range-check, but I'm very concerned about disabling --pointer-primitive-check.

Comment (SaswatPadhi) [reply verification/cbmc/proofs/Makefile-project-defines:43]: So that is a duplicate. The next line also adds `--pointer-primitive-check`.

Comment (feliperodri) [reply verification/cbmc/proofs/Makefile-project-defines:43]: He is not disabling it, there was an extra line. @karkhaz

Comment (karkhaz) [reply verification/cbmc/proofs/Makefile-project-defines:43]: Excuse me, I misread. Thank you!

File: verification/cbmc/proofs/aws_byte_buf_write_be16/cbmc-batch.yaml (hunk 2 above)
Comment (feliperodri) [reply verification/cbmc/proofs/aws_byte_buf_write_be16/cbmc-batch.yaml:8]: @SaswatPadhi could you replace these `cbmc-batch.yaml` files with the following message?
```c
:
  This file marks this directory as containing a CBMC proof. This file
  is automatically clobbered in CI and replaced with parameters for
  running the proof.
```

Comment (SaswatPadhi) [reply verification/cbmc/proofs/aws_byte_buf_write_be16/cbmc-batch.yaml:8]: I was looking at the other proof directories, those that are not modified in this PR, and they don't seem to have this message. Should we make a separate PR and replace all of them?

Comment (feliperodri) [reply verification/cbmc/proofs/aws_byte_buf_write_be16/cbmc-batch.yaml:8]: Yes, let's not pollute this PR with unnecessary changes. We can update the others on a separate one.

Comment (SaswatPadhi) [reply verification/cbmc/proofs/aws_byte_buf_write_be16/cbmc-batch.yaml]: Just pushed all the updated yaml files.

File: verification/cbmc/stubs/memcmp_override.c (hunk 3 above)
Comment (SaswatPadhi) [reply verification/cbmc/stubs/memcmp_override.c:15]: Added a comment. Please check.

**Pull request merged**